homeassistant/components/rfxtrx/__init__.py

```diff
@@ -245,6 +258,20 @@ def handle_receive(event):
         if device_id in device_events:
             hass.bus.fire(EVENT_RFXTRX_EVENT, event_data)
 
+        if config[CONF_AUTOMATIC_ADD]:
+            if device_id not in devices:
+                data = entry.data
+                data[CONF_DEVICES][event_data["data"]] = device_id
```

Comment:
I think you should move this code into the helper function. So he helper function async_add_device takes the event code and device is. Otherwise the helper is rather pointless.

At some point we should probably get back in event loop context quickly from the callback. So we can avoid calling task functions all the time here.

Reply:
Do the helpers run thread-safe? The async_add_device should not run in parallel when multiple events are received in a short time.

Reply:
You should still call it the same way you do with the async threadsafe helper (until we get this in loop context from the start.)

Should be easy to install a minimal event handler that posts a job to event loop that does everything else.

Reply:
I'm not 100% following you, could you clarify it a bit?
You are talking about two approaches right? One is to register to the dispatcher a function to add a device the other is to raise an event from here and install a handler for that?

Reply:
Yes. Forget about the event loop stuff for now. Just let the helper fill out the additional data for the device in the config entry.

Reply:
Ok, thanks for clarification, I'll adjust it.

Reply:
We're not allowed to mutate the entry data directly. It's a mapping proxy type.

We need to copy the entry data before mutating it.